x/feedistribution/keeper/allocation.go

```diff
@@ -0,0 +1,148 @@
+	valBz := val.GetOperator()
+
+	// update current commission
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.EventTypeCommission,
+		sdk.NewAttribute(sdk.AttributeKeyAmount, commission.String()),
+		sdk.NewAttribute(types.EventTypeCommission, val.GetOperator().String()),
+	))
+	currentCommission := k.GetValidatorAccumulatedCommission(ctx, valBz)
+	currentCommission.Commission = currentCommission.Commission.Add(commission...)
+	k.SetValidatorAccumulatedCommission(ctx, valBz, currentCommission)
+	// update current rewards, i.e. the rewards to stakers
+	// if the rewards do not exist it's fine, we will just add to zero.
+	// allocate share tokens to all stakers of this operator.
+	k.AllocateTokensToStakers(ctx, val.GetOperator(), shared, feePool)
+
+	// update outstanding rewards
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.EventTypeRewards,
+		sdk.NewAttribute(sdk.AttributeKeyAmount, commission.String()),
+		sdk.NewAttribute(types.AttributeKeyValidator, val.GetOperator().String()),
+	))
+
+	// ValidatorOutstandingRewards is the rewards of a validator address.
+	outstanding := k.GetValidatorOutstandingRewards(ctx, valBz)
+	outstanding.Rewards = outstanding.Rewards.Add(tokens...)
+	k.SetValidatorOutstandingRewards(ctx, valBz, outstanding)
+	logger.Info("Allocate tokens to validator successfully", "allocated amount is", outstanding.Rewards.String())
+}
+
+func (k Keeper) AllocateTokensToStakers(ctx sdk.Context, operatorAddress sdk.ValAddress, rewardToAllStakers sdk.DecCoins, feePool *types.FeePool) {
+	logger := k.Logger()
+	avsList, err := k.StakingKeeper.GetOptedInAVSForOperator(ctx, operatorAddress.String())
+	if err != nil {
+		ctx.Logger().Error("avs address lists not found; skipping")
+	}
+	stakersPowerMap, curTotoalStakersPowers := make(map[string]math.LegacyDec), math.LegacyNewDec(1)
+	for _, avsAddress := range avsList {
+		avsAssets, err := k.StakingKeeper.GetAVSSupportedAssets(ctx, avsAddress)
+		if err != nil {
+			ctx.Logger().Error("avs address lists not found; skipping")
+		}
+		for assetID := range avsAssets {
+			stakerList, err := k.StakingKeeper.GetStakersByOperator(ctx, operatorAddress.String(), assetID)
+			if err != nil {
+				ctx.Logger().Error("staker lists not found; skipping")
+			}
+			for _, staker := range stakerList.Stakers {
+				if curStakerPower, err := k.StakingKeeper.CalculateUSDValueForStaker(ctx, staker, avsAddress, operatorAddress.Bytes()); err != nil {
+					ctx.Logger().Error("curStakerPower error", err)
+				} else {
+					stakersPowerMap[staker] = curStakerPower
+					curTotoalStakersPowers.Add(curStakerPower)
+				}
+			}
+		}
+	}
+
+	for staker, stakerPower := range stakersPowerMap {
+		powerFraction := stakerPower.QuoTruncate(curTotoalStakersPowers)
+		rewardToSingleStaker := rewardToAllStakers.MulDecTruncate(powerFraction)
+		k.AllocateTokensToSingleStaker(ctx, staker, rewardToSingleStaker)
+		rewardToAllStakers = rewardToAllStakers.Sub(rewardToSingleStaker)
```

Comment:
Safety check that `rewardToAllStakers` isn't negative somehow? Or it is not needed?

Reply:
It'll not be negative. As the deduction is based on the weight ratio of different stakers